File: frontend/src/api/roles.ts
import { apiClient } from './client'

export interface OrganizationRole {
  id: string
  organizationId: string
  name: string
  description: string | null
  permissions: string[] | null
  isSystemRole: boolean
  createdById: string | null
  createdAt: string
  updatedAt: string
}

export interface ProjectRole {
  id: string
  projectId: string
  name: string
  description: string | null
  permissions: string[] | null
  inheritsFromId: string | null
  isSystemRole: boolean
  createdById: string | null
  createdAt: string
  updatedAt: string
}

export interface OrganizationRoleCreate {
  organization_id: string
  name: string
  description?: string
  permissions?: string[]
}

export interface OrganizationRoleUpdate {
  name?: string
  description?: string
  permissions?: string[]
}

export interface ProjectRoleCreate {
  project_id: string
  name: string
  description?: string
  permissions?: string[]
  inherits_from_id?: string
}

export interface ProjectRoleUpdate {
  name?: string
  description?: string
  permissions?: string[]
  inherits_from_id?: string
}

export const rolesApi = {
  // Organization Roles
  listOrganizationRoles: async (organizationId: string): Promise<OrganizationRole[]> => {
    const response = await apiClient.get(`/organizations/${organizationId}/roles`)
    return response.data
  },

  getOrganizationRole: async (organizationId: string, roleId: string): Promise<OrganizationRole> => {
    const response = await apiClient.get(`/organizations/${organizationId}/roles/${roleId}`)
    return response.data
  },

  createOrganizationRole: async (organizationId: string, data: OrganizationRoleCreate): Promise<OrganizationRole> => {
    const response = await apiClient.post(`/organizations/${organizationId}/roles`, data)
    return response.data
  },

  updateOrganizationRole: async (
    organizationId: string,
    roleId: string,
    data: OrganizationRoleUpdate
  ): Promise<OrganizationRole> => {
    const response = await apiClient.put(`/organizations/${organizationId}/roles/${roleId}`, data)
    return response.data
  },

  deleteOrganizationRole: async (organizationId: string, roleId: string): Promise<void> => {
    await apiClient.delete(`/organizations/${organizationId}/roles/${roleId}`)
  },

  // Project Roles
  listProjectRoles: async (projectId: string): Promise<ProjectRole[]> => {
    const response = await apiClient.get(`/projects/${projectId}/roles`)
    return response.data
  },

  getProjectRole: async (projectId: string, roleId: string): Promise<ProjectRole> => {
    const response = await apiClient.get(`/projects/${projectId}/roles/${roleId}`)
    return response.data
  },

  createProjectRole: async (projectId: string, data: ProjectRoleCreate): Promise<ProjectRole> => {
    const response = await apiClient.post(`/projects/${projectId}/roles`, data)
    return response.data
  },

  updateProjectRole: async (projectId: string, roleId: string, data: ProjectRoleUpdate): Promise<ProjectRole> => {
    const response = await apiClient.put(`/projects/${projectId}/roles/${roleId}`, data)
    return response.data
  },

  deleteProjectRole: async (projectId: string, roleId: string): Promise<void> => {
    await apiClient.delete(`/projects/${projectId}/roles/${roleId}`)
  },
}

File: backend/app/api/v1/roles.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.organizations import verify_org_access
from app.core.security import get_current_user
from app.db.session import get_db
from app.models.role import OrganizationRole
from app.models.user import User
from app.schemas.role import OrganizationRoleCreate, OrganizationRoleResponse, OrganizationRoleUpdate

router = APIRouter()


@router.get("/organizations/{org_id}/roles", response_model=list[OrganizationRoleResponse])
async def list_organization_roles(
    org_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List all roles for an organization"""
    await verify_org_access(org_id, current_user, db)

    result = await db.execute(
        select(OrganizationRole)
        .where(OrganizationRole.organization_id == org_id)
        .order_by(OrganizationRole.name)
    )
    roles = result.scalars().all()
    return [OrganizationRoleResponse.model_validate(role) for role in roles]


@router.post("/organizations/{org_id}/roles", response_model=OrganizationRoleResponse, status_code=201)
async def create_organization_role(
    org_id: UUID,
    data: OrganizationRoleCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new custom role for an organization"""
    await verify_org_access(org_id, current_user, db, require_admin=True)

    if data.organization_id != org_id:
        raise HTTPException(status_code=400, detail="Organization ID mismatch")

    existing = await db.execute(
        select(OrganizationRole).where(
            OrganizationRole.organization_id == org_id,
            OrganizationRole.name == data.name
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(status_code=409, detail="Role name already exists in this organization")

    role = OrganizationRole(
        organization_id=org_id,
        name=data.name,
        description=data.description,
        permissions=data.permissions,
        is_system_role=False,
        created_by_id=current_user.id,
    )
    db.add(role)
    await db.flush()

    return OrganizationRoleResponse.model_validate(role)


@router.get("/organizations/{org_id}/roles/{role_id}", response_model=OrganizationRoleResponse)
async def get_organization_role(
    org_id: UUID,
    role_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get a specific organization role by ID"""
    await verify_org_access(org_id, current_user, db)

    result = await db.execute(
        select(OrganizationRole).where(
            OrganizationRole.id == role_id,
            OrganizationRole.organization_id == org_id,
        )
    )
    role = result.scalar_one_or_none()
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")

    return OrganizationRoleResponse.model_validate(role)


@router.put("/organizations/{org_id}/roles/{role_id}", response_model=OrganizationRoleResponse)
async def update_organization_role(
    org_id: UUID,
    role_id: UUID,
    data: OrganizationRoleUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update an existing organization role"""
    await verify_org_access(org_id, current_user, db, require_admin=True)

    result = await db.execute(
        select(OrganizationRole).where(
            OrganizationRole.id == role_id,
            OrganizationRole.organization_id == org_id,
        )
    )
    role = result.scalar_one_or_none()
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")

    if role.is_system_role:
        raise HTTPException(status_code=403, detail="Cannot modify system roles")

    if data.name is not None:
        if data.name != role.name:
            dup = await db.execute(
                select(OrganizationRole).where(
                    OrganizationRole.organization_id == org_id,
                    OrganizationRole.name == data.name,
                    OrganizationRole.id != role_id
                )
            )
            if dup.scalar_one_or_none():
                raise HTTPException(status_code=409, detail="Role name already exists in this organization")
            role.name = data.name

    if data.description is not None:
        role.description = data.description
    if data.permissions is not None:
        role.permissions = data.permissions

    await db.flush()
    return OrganizationRoleResponse.model_validate(role)


@router.delete("/organizations/{org_id}/roles/{role_id}")
async def delete_organization_role(
    org_id: UUID,
    role_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete an organization role"""
    await verify_org_access(org_id, current_user, db, require_admin=True)

    result = await db.execute(
        select(OrganizationRole).where(
            OrganizationRole.id == role_id,
            OrganizationRole.organization_id == org_id,
        )
    )
    role = result.scalar_one_or_none()
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")

    if role.is_system_role:
        raise HTTPException(status_code=403, detail="Cannot delete system roles")

    await db.delete(role)
    await db.flush()

    return {"message": "Role deleted"}
